src/ir/stack-utils.cpp

```diff
@@ -0,0 +1,278 @@
+  processBlock([&](Expression* expr, StackSignature sig) {
+    assert((sig.params.size() <= values.size() || lastUnreachable) &&
+           "Block inputs not yet supported");
+
+    // Unreachable instructions consume all available values
+    size_t consumed = sig.unreachable
+                        ? std::max(values.size(), sig.params.size())
+                        : sig.params.size();
+
+    // We previously calculated how many values unreachable instructions produce
+    size_t produced =
+      sig.unreachable ? producedByUnreachable[expr] : sig.results.size();
+
+    srcs[expr] = std::vector<Location>(consumed);
+    dests[expr] = std::vector<Location>(produced);
+
+    // Consume values
+    assert(sig.params.size() <= consumed);
+    size_t unreachableConsumed =
+      consumed > values.size() ? consumed - values.size() : 0;
+    for (Index i = 0; i < consumed; ++i) {
+      if (i < unreachableConsumed) {
+        // This value comes from the polymorphic stack of the last unreachable
+        // because the stack did not have enough values to satisfy this
+        // instruction.
+        assert(consumed == sig.params.size());
+        assert(lastUnreachable);
+        assert(producedByUnreachable[lastUnreachable] >= unreachableConsumed);
+        Index destIndex =
+          producedByUnreachable[lastUnreachable] - unreachableConsumed + i;
+        Type type = sig.params[i];
+        srcs[expr][i] = {lastUnreachable, destIndex, type, true};
+        dests[lastUnreachable][destIndex] = {expr, i, type, false};
+      } else {
+        // A normal value from the value stack
+        bool unreachable = sig.params.size() + i < consumed;
+        auto& src = values[values.size() + i - consumed];
+        srcs[expr][i] = {src.expr, src.index, src.type, false};
+        dests[src.expr][src.index] = {expr, i, src.type, unreachable};
+      }
+    }
+    // Update available values
+    if (unreachableConsumed) {
+      producedByUnreachable[lastUnreachable] -= unreachableConsumed;
+      values.resize(0);
+    } else {
+      values.resize(values.size() - consumed);
+    }
+    // Produce values
+    for (Index i = 0; i < sig.results.size(); ++i) {
+      values.push_back({expr, i, sig.results[i], false});
+    }
+    // Update the last unreachable instruction
+    if (sig.unreachable) {
+      assert(producedByUnreachable[lastUnreachable] == 0);
+      lastUnreachable = expr;
+    }
+  });
+}
+
+StackSignature StackFlow::getSignature(Expression* expr) {
+  auto exprSrcs = srcs.find(expr);
+  auto exprDests = dests.find(expr);
+  assert(exprSrcs != srcs.end() && exprDests != dests.end());
+  std::vector<Type> params, results;
+  for (auto& src : exprSrcs->second) {
+    params.push_back(src.type);
+  }
+  for (auto& dest : exprDests->second) {
+    results.push_back(dest.type);
+  }
+  bool unreachable = expr->type == Type::unreachable;
+  return StackSignature(Type(params), Type(results), unreachable);
```

Comment:
This is the same question I asked somewhere above, but this seems to be imply that an expression's `StackSignature` can be different when it is computed independently and when it is in the middle of a block. For example, if you compute `StackSignature` of an `unreachable` instruction independently it will be `{params=none, results=none, unreachable=true}`. But when it is in the middle of a block, its `params` and `results` computed here can be different depending on its polymorphic behavior. So is it correct to think that `StackSignature` for an expression changes depending on how it is computed?

Reply:
Yes, a polymorphic instruction can have different StackSignatures in different contexts. The `StackSignature(Expression*)` constructor and the iterator constructor both construct the "minimal" signature for a polymorphic instruction or sequence of instructions, but prepending arbitrary types to either the params or dests of that signature will give you more stack signatures that could describe the same instruction or sequence in different contexts. `StackFlow` calculates _exact_ stack signatures for unreachable instructions given their context.